File: javascript/src/2.两数相加.js
/*
 * @lc app=leetcode.cn id=2 lang=javascript
 *
 * [2] 两数相加
 *
 * https://leetcode-cn.com/problems/add-two-numbers/description/
 *
 * algorithms
 * Medium (36.50%)
 * Likes:    3843
 * Dislikes: 0
 * Total Accepted:    314.3K
 * Total Submissions: 861K
 * Testcase Example:  '[2,4,3]\n[5,6,4]'
 *
 * 给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，并且它们的每个节点只能存储 一位 数字。
 * 
 * 如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。
 * 
 * 您可以假设除了数字 0 之外，这两个数都不会以 0 开头。
 * 
 * 示例：
 * 
 * 输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
 * 输出：7 -> 0 -> 8
 * 原因：342 + 465 = 807
 * 
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    return solution1(l1, l2);
};

var solution1 = function (l1, l2) {
    var value1 = getNumber(l1)
    var value2 = getNumber(l2)
    var value = value1 + value2;

    var array = []
    while (value > 0) {
        array.push(value % 10)
        value = parseInt(value / 10)
    }
}

var getNumber = function (l1) {
    var stack1 = [];
    var value1 = 0;
    do {
        stack1.push(l1.val)
        l1 = l1.next
    } while (l1 != null)

    while (stack1.length > 0) {
        value1 = stack1.pop() + value1 * 10;
    }

    return value1;
}
// @lc code=end

File: java/src/2.两数相加.java
/*
 * @lc app=leetcode.cn id=2 lang=java
 *
 * [2] 两数相加
 *
 * https://leetcode-cn.com/problems/add-two-numbers/description/
 *
 * algorithms
 * Medium (40.28%)
 * Likes:    6336
 * Dislikes: 0
 * Total Accepted:    873.8K
 * Total Submissions: 2.2M
 * Testcase Example:  '[2,4,3]\n[5,6,4]'
 *
 * 给你两个 非空 的链表，表示两个非负的整数。它们每位数字都是按照 逆序 的方式存储的，并且每个节点只能存储 一位 数字。
 * 
 * 请你将两个数相加，并以相同形式返回一个表示和的链表。
 * 
 * 你可以假设除了数字 0 之外，这两个数都不会以 0 开头。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：l1 = [2,4,3], l2 = [5,6,4]
 * 输出：[7,0,8]
 * 解释：342 + 465 = 807.
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：l1 = [0], l2 = [0]
 * 输出：[0]
 * 
 * 
 * 示例 3：
 * 
 * 
 * 输入：l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
 * 输出：[8,9,9,9,0,0,0,1]
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 每个链表中的节点数在范围 [1, 100] 内
 * 0 
 * 题目数据保证列表表示的数字不含前导零
 * 
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int length1 = getLength(l1);
        int length2 = getLength(l2);
        
        ListNode p = l1,q = l2;
        // 短的后面补零
        if(length1>length2){
           addZero(l2, length1-length2);
        }else if(length1<length2){
           addZero(l1, length2-length1);
        }
        ListNode ans = new ListNode();
        ListNode k = ans;
        int plus = 0;
        while(p!=null&&q!=null){
            int number = plus+p.val+q.val;
            if(number>=10){
                number = number%10;
                plus = 1;
            }else{
                plus = 0;
            }
            p = p.next;
            q = q.next;

            ListNode node = new ListNode(number);
            k.next = node;
            k = k.next;
        }
        // 检查最后的进位
        if(plus>0){
            k.next = new ListNode(1);
        }
        return ans.next;
    }

    private ListNode addZero(ListNode node, int count){
        ListNode ans = node;
        while(node.next!=null){
            node= node.next;
        }
        while(count>0){
            node.next = new ListNode(0);
            node = node.next;
            count--;
        }
        return ans;
    }

    private int getLength(ListNode node){
        int ans = 0;
        while(node!=null){
            node = node.next;
            ans++; 
        }
        return ans;
    }
}
// @lc code=end

